### PR TITLE
content type support

### LINKED ISSUE #186: Can't customize my own datacodec
From the source code, it seems "io.scalecube.services.annotations.ContentType" is not supported yet.

### PR DESCRIPTION
**Pull request by segabriel**
resolves #186

### CODE REVIEW

File: services/src/test/java/io/scalecube/services/routings/sut/WeightedRandomRouter.java [hunk 1]
```diff
@@ -12,7 +12,7 @@
   @Override
   public Optional<ServiceReference> route(ServiceRegistry serviceRegistry, ServiceMessage request) {
     RandomCollection<ServiceReference> weightedRandom = new RandomCollection<>();
-    serviceRegistry.lookupService(request.qualifier())
+    serviceRegistry.lookupService(request.qualifier(), request.dataFormat())
```
Comment (ronenhamias) [inline services/src/test/java/io/scalecube/services/routings/sut/WeightedRandomRouter.java]: how about changing the interface to serviceRegistry.serviceLookup(request) ?
WDYT?

File: services-codec/src/main/java/io/scalecube/services/codec/DataCodec.java [hunk 2]
```diff
@@ -5,22 +5,39 @@
...
+  }
 
   static DataCodec getInstance(String contentType) {
-    return INSTANCES.computeIfAbsent(contentType, DataCodec::loadInstance);
+    if (contentType == null) {
+      throw new IllegalArgumentException("contentType not specified");
+    }
+    DataCodec dataCodec = INSTANCES.get(contentType);
+    if (dataCodec == null) {
+      throw new IllegalStateException("DataCodec for '" + contentType + "' not configured");
```
Comment (artem-v) [inline services-codec/src/main/java/io/scalecube/services/codec/DataCodec.java]: Make it `IllegalArgumentException`, this is not `ISE`, our state is good, it's probably parameter `contentType` is wrong

**Review by artem-v — CHANGES_REQUESTED** [review]

File: services-codec/src/main/java/io/scalecube/services/codec/DataCodec.java [hunk 3]
```diff
@@ -5,22 +5,39 @@
...
+    }
+    return dataCodec;
   }
 
-  static DataCodec loadInstance(String contentType) {
-    Optional<DataCodec> result = ServiceLoaderUtil.findFirst(DataCodec.class,
-        codec -> codec.contentType().equalsIgnoreCase(contentType));
-    return result.orElseThrow(() -> new IllegalStateException("DataCodec not configured"));
+  static DataCodec getInstance(String contentType, String defaultContentType) {
```
Comment (artem-v) [inline services-codec/src/main/java/io/scalecube/services/codec/DataCodec.java]: Cool method but it left client confusing when he sees another one defined above. I think it makes sense to rename this one to `getInstanceOrDefault` , and that one which throws `getInstanceOrThrow`. WDYT?

File: services/src/main/java/io/scalecube/services/registry/ServiceRegistryImpl.java [hunk 4]
```diff
@@ -38,9 +38,17 @@
   }
 
   @Override
-  public List<ServiceReference> lookupService(String qualifier) {
+  public List<ServiceReference> lookupService(String qualifier, String contentType) {
     List<ServiceReference> result = referencesByQualifier.get(qualifier);
-    return result != null ? Collections.unmodifiableList(result) : Collections.emptyList();
+    if (result != null) {
+      if (contentType != null) {
```
Comment (artem-v) [inline services/src/main/java/io/scalecube/services/registry/ServiceRegistryImpl.java]: if content type is null, don't we have to take JSON as default one data format?

**Review by artem-v — CHANGES_REQUESTED** [review]

File: services-api/src/main/java/io/scalecube/services/ServiceLoaderUtil.java [hunk 5]
```diff
@@ -12,14 +12,53 @@ private ServiceLoaderUtil() {
...
+   * creates their instances.
+   *
+   * @param aClass    service type
+   * @param predicate service type predicate
+   * @return implementations' stream of the given service type
+   */
+  public static <T> Stream<T> findAll(Class<T> aClass, Predicate<? super T> predicate) {
+    ServiceLoader<T> load = ServiceLoader.load(aClass);
+    return StreamSupport.stream(load.spliterator(), false)
+        .filter(predicate);
```
Comment (artem-v) [inline services-api/src/main/java/io/scalecube/services/ServiceLoaderUtil.java]: Would you mind if I ask you to put this `.filter(predicate`  at the line 61. Thanks )

File: services-api/src/main/java/io/scalecube/services/ServiceLoaderUtil.java [hunk 6]
```diff
@@ -12,14 +12,53 @@ private ServiceLoaderUtil() {
     // Do not instantiate
   }
 
+  /**
+   * Finds the first implementation of the given service type and creates its instance.
+   *
+   * @param aClass service type
+   * @return the first implementation of the given service type
+   */
   public static <T> Optional<T> findFirstMatched(Class<T> aClass) {
```
Comment (artem-v) [inline services-api/src/main/java/io/scalecube/services/ServiceLoaderUtil.java]: It makes sense to rename this dude to `findFirst`

File: services-codec/src/main/java/io/scalecube/services/codec/DataCodec.java (hunk 3 above)
Comment (segabriel) [reply services-codec/src/main/java/io/scalecube/services/codec/DataCodec.java]: yeah OK

**Review by artem-v — CHANGES_REQUESTED** [review]

File: services/src/main/java/io/scalecube/services/registry/ServiceRegistryImpl.java [hunk 7]
```diff
@@ -38,9 +38,17 @@
...
+        result = result.stream().filter(ref -> ref.contentTypes().contains(contentType)).collect(Collectors.toList());
+      }
+      result = Collections.unmodifiableList(result);
```
Comment (artem-v) [inline services/src/main/java/io/scalecube/services/registry/ServiceRegistryImpl.java]: Let's not create `unmodifiableList` on critical path function

**Review by artem-v — CHANGES_REQUESTED** [review]

File: services-api/src/main/java/io/scalecube/services/ServiceEndpoint.java [hunk 8]
```diff
@@ -40,6 +44,10 @@ public int port() {
     return port;
   }
 
+  public Set<String> contentTypes() {
+    return contentTypes;
```
Comment (dmytro-lazebnyi) [inline services-api/src/main/java/io/scalecube/services/ServiceEndpoint.java:48]: Let's return `Collections.unmodifiableSet`.

File: services-api/src/main/java/io/scalecube/services/ServiceReference.java [hunk 9]
```diff
@@ -63,8 +64,8 @@ public String namespace() {
     return namespace;
   }
 
-  public String contentType() {
-    return contentType;
+  public Set<String> contentTypes() {
+    return contentTypes;
```
Comment (dmytro-lazebnyi) [inline services-api/src/main/java/io/scalecube/services/ServiceReference.java:64]: Let's return `Collections.unmodifiableSet`.

File: services/src/test/java/io/scalecube/services/routings/sut/WeightedRandomRouter.java (hunk 1 above)
Comment (segabriel) [reply services/src/test/java/io/scalecube/services/routings/sut/WeightedRandomRouter.java]: I agree with You, but my teammates want this component to be decoupling from `ServiceMessage`.

**Review by artem-v — DISMISSED** [review]

File: services-api/src/main/java/io/scalecube/services/ServiceReference.java (hunk 9 above)
Comment (segabriel) [reply services-api/src/main/java/io/scalecube/services/ServiceReference.java:64]: I did this `ServiceReference` as an immutable and add `Collections.unmodifiableSet()` for content-types in constructor.

**Review by artem-v — APPROVED** [review]

**Review by artem-v — APPROVED** [review]

**Pull request merged**